lib/widgets/set_status.dart

```diff
@@ -0,0 +1,339 @@
+    return MaterialAccountWidgetRoute(context: context,
+      page: SetStatusPage(oldStatus: oldStatus));
+  }
+
+  @override
+  State<SetStatusPage> createState() => _SetStatusPageState();
+}
+
+class _SetStatusPageState extends State<SetStatusPage> {
+  late final TextEditingController statusTextController;
+  late final ValueNotifier<UserStatusChange> statusChange;
+
+  UserStatus get oldStatus => widget.oldStatus;
+  UserStatus get newStatus => statusChange.value.apply(widget.oldStatus);
+
+  @override
+  void initState() {
+    super.initState();
+    statusTextController = TextEditingController(text: oldStatus.text)
+      ..addListener(() {
+        final trimmedValue = statusTextController.text.trim();
+        final text = trimmedValue.isNotEmpty ? trimmedValue : null;
+
+        // Ignore updating [statusChange] for the additional updates with the
+        // same value from TextField. For example, when a character is deleted.
+        if (text == newStatus.text) return;
+
+        statusChange.value = statusChange.value.copyWith(
+          text: toOption(recent: text, old: oldStatus.text),
+        );
+      });
+    statusChange =
+      ValueNotifier(UserStatusChange(text: OptionNone(), emoji: OptionNone()))
+        ..addListener(() {
+          final text = statusChange.value.text.or(oldStatus.text) ?? '';
+
+          // Ignore updating the status text field if it already has the same
+          // text. It can happen in the following cases:
+          //   1. Only the emoji is changed.
+          //   2. The same status is chosen consecutively from the suggested
+          //      statuses list.
+          //   3. This listener is called as a result of the change in status
+          //      text field.
+          if (text == statusTextController.text) return;
+
+          statusTextController.text = text;
+        });
+  }
+
+  @override
+  void dispose() {
+    statusTextController.dispose();
+    statusChange.dispose();
+    super.dispose();
+  }
+
+  List<UserStatus> statusSuggestions(BuildContext context) {
+    final store = PerAccountStoreWidget.of(context);
+    final localizations = ZulipLocalizations.of(context);
+
+    final values = [
+      (localizations.userStatusBusy,            '1f6e0', 'working_on_it'),
+      (localizations.userStatusInAMeeting,      '1f4c5', 'calendar'),
+      (localizations.userStatusCommuting,       '1f68c', 'bus'),
+      (localizations.userStatusOutSick,         '1f912', 'sick'),
+      (localizations.userStatusVacationing,     '1f334', 'palm_tree'),
+      (localizations.userStatusWorkingRemotely, '1f3e0', 'house'),
+      (localizations.userStatusAtTheOffice,     '1f3e2', 'office'),
+    ];
+
+    return [
+      for (final (statusText, emojiCode, fallbackEmojiName) in values)
+        UserStatus(
+          text: statusText,
+          emoji: StatusEmoji(
+            emojiName: store.allEmojiCandidates()
+                         .firstWhereOrNull((e) => e.emojiCode == emojiCode)
```

Comment:
Hmm, this iterates through the whole list of all known emoji, repeatedly for each of the suggestions we want to offer. Let's avoid doing that. :slightly_smiling_face: 

Instead, have a prep commit add a little method to EmojiStore to offer the API we need in order to do this efficiently. I think that can look like `String? getUnicodeEmojiNameByCode(String emojiCode)`.